### PR TITLE
feat: convert to or modify multisig account added

### PR DESCRIPTION
**Pull request by yilmazbahadir**
`symbol-bootstrap modifyMultisig`  command is added. It's useful to convert main account into a multisig account or to modify the multisig account structure. Takes the following parameters, the user will be prompted for the unset parameters.

```
-A, --addressAdditions=addressAdditions              Cosignatory accounts addresses to be added (separated by a
                                                       comma).

  -D, --addressDeletions=addressDeletions              Cosignatory accounts addresses to be removed (separated by a
                                                       comma).

  -a, --minApprovalDelta=minApprovalDelta              Number of signatures needed to approve a transaction. If the
                                                       account is already multisig, enter the number of cosignatories to
                                                       add or remove.

  -r, --minRemovalDelta=minRemovalDelta                Number of signatures needed to remove a cosignatory. If the
                                                       account is already multisig, enter the number of cosignatories to
                                                       add or remove.
```

### CODE REVIEW

File: src/service/AnnounceService.ts [hunk 1]
```diff
@@ -358,18 +358,39 @@ export class AnnounceService {
...
-                            providedMaxFee,
-                            minFeeMultiplier,
-                            generationHash,
-                            currency,
-                            transactionService,
-                            listener,
-                            ready,
-                            nodeAccount.name,
-                        );
+                        if (transactions[0] instanceof MultisigAccountModificationTransaction) {
```
Comment (rg911) [inline src/service/AnnounceService.ts:361]: would it be better to use the transactionType instead?

File: src/service/AnnounceService.ts [hunk 2]
```diff
@@ -358,18 +358,39 @@ export class AnnounceService {
...
+                                generationHash,
+                                currency,
+                                transactionService,
+                                listener,
+                                ready,
+                                nodeAccount.name,
+                            );
+                        }
                     } else {
                         await this.announceAggregateComplete(
```
Comment (rg911) [inline src/service/AnnounceService.ts:395]: does the modifyMultisig tx needs to be announced as a bonded transaction? Using transactions.length might not always be safe. What if modifyMultisig is part of the transactions array? what about using array.filter?

File: src/service/ModifyMultisigService.ts [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+};
+
+const logger: Logger = LoggerFactory.getLogger(LogType.System);
+
+export class ModifyMultisigService implements TransactionFactory {
+    public static readonly defaultParams: ModifyMultisigParams = {
+        target: BootstrapUtils.defaultTargetFolder,
+        useKnownRestGateways: false,
+        ready: false,
+        url: 'http://localhost:3000',
```
Comment (rg911) [inline src/service/ModifyMultisigService.ts:52]: default to hardcoded http?

File: src/service/AnnounceService.ts (hunk 1 above)
Comment (yilmazbahadir) [reply src/service/AnnounceService.ts:361]: why do you think so?

Comment (rg911) [reply src/service/AnnounceService.ts:361]: I mean you can use instanceof, but the SDK has already got a transactionType enum. could that be neat using the type?

File: src/service/ModifyMultisigService.ts (hunk 3 above)
Comment (yilmazbahadir) [reply src/service/ModifyMultisigService.ts:52]: yes and default to localhost :)
Copied from: https://github.com/symbol/symbol-bootstrap/blob/56239efa3100116ea294c570ddab317160e0e00b/src/service/LinkService.ts#L77
I think it's fine for this PR, the node list loading logic will be changed after @fboucquez 's (soon to be raised) PR.

File: src/commands/modifyMultisig.ts [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+            char: 'r',
+        }),
+        minApprovalDelta: flags.integer({
+            description:
+                'Delta of signatures needed to approve a transaction. ' +
+                '0 means no change, a positive(+) number means increment and a negative(-) number means decrement to the actual value.',
+            char: 'a',
+        }),
+        addressAdditions: flags.string({
+            description: 'Cosignatory accounts addresses to be added (separated by a comma).',
```
Comment (fboucquez) [inline src/commands/modifyMultisig.ts:46]: have you tried `multiple: true`, cli handles lists of strings. Does it work nicely?

File: src/service/ModifyMultisigService.ts (hunk 3 above)
Comment (fboucquez) [reply src/service/ModifyMultisigService.ts:52]: Yes, it's fine, default url is when you are running the link/modify command on the same box that you have launched the node. Most people would either provide an `--url` or use `--useKnownRestGateways`

**Review by fboucquez — CHANGES_REQUESTED** [review]
Please also run `npm run style:fix` to update the index.ts and docs

I have done the following testing:

Multisig M, cosigners A,B,C , ServiceOperatorAccount S. It looks really nice:

1) Convert to multisig M using M private key. It creates a bonded where All A,B,C need to cosign.
2) Modify mutlisg M using A cosigner private key. It creates a bonded where some of B,C to cosign
3) Convert to multisig  M using S private key. It creates a bonded where all M, A, B, C need to cosign (wallet warning shown as expected)
4) Modify multisig M using S private key. It creates a bonded where some A, B, C need to cosign (wallet warning shown as expected)

Could the command ask (optionally) for cosigners private keys. Use case, I'm a regular user that wants to convert my main account to multisig for security reasons. I could provide my cosigners private keys on the spot without bonded, using full aggregate complete. This could be an improvement though, the service providers should never know any cosigner private key.

File: src/commands/modifyMultisig.ts (hunk 4 above)
Comment (yilmazbahadir) [reply src/commands/modifyMultisig.ts:46]: Yes, tried multiple addresses separated by comma, works fine 👍

**Comment by yilmazbahadir** [comment]
> Please also run `npm run style:fix` to update the index.ts and docs
> 
> I have done the following testing:
> 
> Multisig M, cosigners A,B,C , ServiceOperatorAccount S. It looks really nice:
> 
> 1. Convert to multisig M using M private key. It creates a bonded where All A,B,C need to cosign.
> 2. Modify mutlisg M using A cosigner private key. It creates a bonded where some of B,C to cosign
> 3. Convert to multisig  M using S private key. It creates a bonded where all M, A, B, C need to cosign (wallet warning shown as expected)
> 4. Modify multisig M using S private key. It creates a bonded where some A, B, C need to cosign (wallet warning shown as expected)
> 
> Could the command ask (optionally) for cosigners private keys. Use case, I'm a regular user that wants to convert my main account to multisig for security reasons. I could provide my cosigners private keys on the spot without bonded, using full aggregate complete. This could be an improvement though, the service providers should never know any cosigner private key.

Great! Will run the `style:fix` script. Let's implement this improvement with a different PR, created https://github.com/symbol/symbol-bootstrap/issues/308.
Currently writing some unit tests and also adding some parameter sanity checks, will be updating the PR shortly.

File: src/commands/modifyMultisig.ts (hunk 4 above)
Comment (fboucquez) [reply src/commands/modifyMultisig.ts:46]: I mean, oclif `multiple: true` field in the flags.string. It users cli style multiple values

Comment (fboucquez) [reply src/commands/modifyMultisig.ts:46]: ignore, --addressAdditions A,B,C its easier than --addressAdditions A --addressAdditions B --addressAdditions C

does  `--addressAdditions` (empty) work? empty list no prompt.

Comment (yilmazbahadir) [reply src/commands/modifyMultisig.ts:46]: I think so
>ignore, --addressAdditions A,B,C is easier than --addressAdditions A --addressAdditions B --addressAdditions C

No it doesn't, it gives => No Error: Flag --addressAdditions expects a value
> does --addressAdditions (empty) work? empty list no prompt.

If we need to support no-prompt command as well, I think we have 2 options here:
1. Remove all the prompts, use only the flag values
2. Add the following flags --no-addressAdditions, --no-addressDeletions 
    and the command will look like
    `symbol-bootstrap modifyMultisig --no-addressAdditions --no-addressDeletions --minApprovalDelta=1 --minRemovalDelta=0`
I'd probably vote for 2.

Comment (yilmazbahadir) [reply src/commands/modifyMultisig.ts:46]: Updated the code as it will not prompt for the following.
`--addressAdditions='' --addressDeletions='' `

**Pull request merged**